feat(www): adds documentation for useQueries hook (#3419)
trpc/trpc: www/docs/reactjs/useQueries.mdx

---
id: usequeries
title: useQueries()
sidebar_label: useQueries()
slug: /use-queries
---

> The useQueries hook is used to fetch multiple queries using one hook call. This hook is an excellent choice in scenarios where you fetch numerous queries in one component as it can lead to a cleaner and more readable code.

Reviewer: Is this really what `useQueries` is for? cc @TkDodo, I'm unsure even after reading the docs page

Reviewer 2: The main use-case would be for a variable amount of queries, usually of the same type. So, fetch a list of todos, then fetch all details for all todos in a `useQueries` where you `map` over the result of the first query.

having one `useQueries` where you fetch two different resources with different types is certainly possible, but there isn't much advantage over calling `useQuery` twice I believe, unless you use `suspense`, because `useQueries` can trigger suspense in parallel, which `useQuery` cannot (it would waterfall).

Reviewer 3: @ixahmedxi could we get the examples to reflect this?

i.e.

```ts
trpc.useQueries(t => postIds.map(id => t.post.byId({ id })))
```

## Usage

The useQueries hook is the same as that of [@tanstack/query useQueries](https://tanstack.com/query/v4/docs/react/reference/useQueries). The only difference is that you pass in a function that returns an array of queries instead of an array of queries inside an object parameter.

```tsx
const Component = () => {
  const [post, greeting] = trpc.useQueries((t) => [
    t.post.byId({ id: '1' }),
    t.greeting({ text: 'world' }),

>   ]);

Reviewer: like Alex mentioned on Discord - twoslash would be great here

  return (
    <div>
      <h1>{post.data.title}</h1>
      <p>{greeting.data.message}</p>
    </div>
  );
};
```

### Providing options to individual queries

You can also pass in any normal query options to the second parameter of any of the query calls in the array such as `enabled`, `suspense`, `refetchOnWindowFocus`...etc. For a complete overview of all the available options, see the [tanstack useQuery](https://tanstack.com/query/v4/docs/react/reference/useQuery) documentation.

```tsx
const Component = () => {
  const [post, greeting] = trpc.useQueries((t) => [
    t.post.byId({ id: '1' }, { enabled: false }),
    t.greeting({ text: 'world' }),
  ]);

  const onButtonClick = () => {
    post.refetch();
  };

  return (
    <div>
      <h1>{post.data && post.data.title}</h1>
      <p>{greeting.data.message}</p>
      <button onClick={onButtonClick}>Click to fetch</button>
    </div>
  );
};
```

> ## Options
> 
> ### `(t) => [t.query1(), t.query2()]`
> 
> The first parameter of the useQueries hook is a function that contains a `t` object with all of the queries you can fetch. The function should return an array of `t` query calls with their inputs and options.
> 
> ```tsx
> const [post, greeting] = trpc.useQueries((t) => [
>   t.post.byId({ id: '1' }),
>   t.greeting({ text: 'world' }),
> ]);
> ```

Reviewer: ```suggestion

```

I don't think we need to describe the interface, it's difficult to understand just by reading. The example speaks more than words.

### Context

> You can also pass in an optional context if you would not like to use the default context of tRPC.

Reviewer: ```suggestion
You can also pass in an optional React Query context to override the default.
```
